core/src/test/java/org/bitcoinj/wallet/WalletTest.java

```diff
@@ -2737,6 +2739,108 @@ public void higherThanDefaultFee() throws InsufficientMoneyException {
+
+        /** ASSERT */
+        isValidSignedTx(deserialisedTx, hash -> wallet.getTransaction(hash));
+    }
+
+    @Test
+    public void whenValidDeserialisedTx_Then_OutputsShouldBeUnchanged() throws InsufficientMoneyException {
+
+        /** ARRANGE */
+        sendMoneyToWallet(AbstractBlockChain.NewBlockType.BEST_CHAIN, valueOf(3, 0));
+        Transaction origTxn = new Transaction();
+        origTxn.addOutput(valueOf(0, 50), OTHER_ADDRESS);
+        SendRequest firstSendRequest = SendRequest.forTx(origTxn);
+        wallet.completeTx(firstSendRequest);
+        byte[] txBytes = firstSendRequest.tx.bitcoinSerialize();
+        Transaction deserialisedTx = Transaction.read(ByteBuffer.wrap(txBytes));
+        SendRequest deserialisedSendRequest = SendRequest.forTx(deserialisedTx);
+
+        /** ACT */
+        wallet.completeTx(deserialisedSendRequest);
+
+        /** ASSERT */
+        List<TransactionInput> inputs = deserialisedSendRequest.tx.getInputs();
+        assertEquals(1, inputs.size());
+        TransactionInput originalInput = firstSendRequest.tx.getInput(0);
+        TransactionInput actualInput = inputs.get(0);
+        assertEquals(originalInput.getOutpoint().hash(), actualInput.getOutpoint().hash());
+        assertEquals(originalInput.getOutpoint().index(), actualInput.getOutpoint().index());
+        isValidSignedTx(deserialisedSendRequest.tx, hash -> wallet.getTransaction(hash));
+    }
+
+    @Test
+    public void GIVEN_sendrequest_with_input_THEN_dont_reuse_same_input() throws Exception {
+
+        /** ARRANGE */
+        StoredBlock block = new StoredBlock(makeSolvedTestBlock(blockStore, OTHER_ADDRESS), BigInteger.ONE, 1);
+        Transaction tx1 = createFakeTx(TESTNET.network(), COIN, myAddress);
+        wallet.receiveFromBlock(tx1, block, AbstractBlockChain.NewBlockType.BEST_CHAIN, 0);
+        Transaction tx2 = createFakeTx(TESTNET.network(), COIN, myAddress);
+        wallet.receiveFromBlock(tx2, block, AbstractBlockChain.NewBlockType.BEST_CHAIN, 0);
+
+        // SendRequest using tx1 as input
+        SendRequest request1 = SendRequest.to(OTHER_ADDRESS, COIN.multiply(1));
+        request1.tx.addInput(new TransactionInput(request1.tx, new byte[] {}, new TransactionOutPoint(0, tx1.getTxId())));
+
+        // SendRequest using tx2 as input
+        SendRequest request2 = SendRequest.to(OTHER_ADDRESS, COIN.multiply(1));
+        request2.tx.addInput(new TransactionInput(request2.tx, new byte[] {}, new TransactionOutPoint(0, tx2.getTxId())));
+
+        /** ACT */
+        wallet.completeTx(request1);
+        wallet.completeTx(request2);
+
+        /** ASSERT */
+        Function<List<TransactionInput>, Set<String>> uniqueInputs = txInputs -> txInputs.stream().map(txInput ->{
+            TransactionOutPoint outpoint = txInput.getOutpoint();
+            String result = outpoint.hash().toString() + outpoint.index();
+            return result;
+        }).collect(Collectors.toSet());
```

Comment:
I think this would be simpler and more readable as a helper method. I also think the type of the `Set` could be `Set<TransactionOutPoint>` and rely on `TransactionOutPoint::equals`. Something like:
```
Set<TransactionOutPoint> uniqueOutPoints(List<TransactionInput>) {
...
}
```